File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-tailwind-visibility-at-breakpoints.ts
import { TAILWIND_BREAKPOINT_NAMES } from "../constants/tailwind.js";
import { getTailwindVisibilityEffect } from "./get-tailwind-visibility-effect.js";
import type { TailwindVisibilityEffect } from "./get-tailwind-visibility-effect.js";
import { parseTailwindClassNameToken } from "./parse-tailwind-class-name-token.js";
import type { TailwindClassNameToken } from "./parse-tailwind-class-name-token.js";
import { splitTailwindClassName } from "./split-tailwind-class-name.js";

interface TailwindResponsiveVariantScope {
  maximumBreakpointIndex: number;
  minimumBreakpointIndex: number;
  specificity: number;
}

interface TailwindScopedVisibilityEffect {
  effect: TailwindVisibilityEffect;
  scope: TailwindResponsiveVariantScope;
  token: TailwindClassNameToken;
}

const getResponsiveVariantScope = (
  variants: ReadonlyArray<string>,
): TailwindResponsiveVariantScope | null | undefined => {
  let minimumBreakpointIndex = 0;
  let maximumBreakpointIndex = TAILWIND_BREAKPOINT_NAMES.length;
  for (const variant of variants) {
    const minimumVariantIndex = TAILWIND_BREAKPOINT_NAMES.indexOf(variant);
    if (minimumVariantIndex > 0) {
      minimumBreakpointIndex = Math.max(minimumBreakpointIndex, minimumVariantIndex);
      continue;
    }
    if (variant.startsWith("max-")) {
      const maximumVariantIndex = TAILWIND_BREAKPOINT_NAMES.indexOf(variant.slice("max-".length));
      if (maximumVariantIndex > 0) {
        maximumBreakpointIndex = Math.min(maximumBreakpointIndex, maximumVariantIndex);
        continue;
      }
    }
    return null;
  }
  return {
    maximumBreakpointIndex,
    minimumBreakpointIndex,
    specificity: variants.length,
  };
};

const resolveVisibilityProperty = (
  scopedEffects: ReadonlyArray<TailwindScopedVisibilityEffect>,
  breakpointIndex: number,
  propertyName: TailwindVisibilityEffect["propertyName"],
): boolean | null => {
  const applicableEffects = scopedEffects.filter(
    ({ effect, scope }) =>
      effect.propertyName === propertyName &&
      breakpointIndex >= scope.minimumBreakpointIndex &&
      breakpointIndex < scope.maximumBreakpointIndex,
  );
  if (applicableEffects.length === 0) return true;
  const hasImportantEffect = applicableEffects.some(({ token }) => token.isImportant);
  const highestImportanceEffects = hasImportantEffect
    ? applicableEffects.filter(({ token }) => token.isImportant)
    : applicableEffects;
  const maximumSpecificity = Math.max(
    ...highestImportanceEffects.map(({ scope }) => scope.specificity),
  );
  const highestSpecificityEffects = highestImportanceEffects.filter(
    ({ scope }) => scope.specificity === maximumSpecificity,
  );
  const maximumMinimumBreakpoint = Math.max(
    ...highestSpecificityEffects.map(({ scope }) => scope.minimumBreakpointIndex),
  );
  const latestMinimumEffects = highestSpecificityEffects.filter(
    ({ scope }) => scope.minimumBreakpointIndex === maximumMinimumBreakpoint,
  );
  const minimumMaximumBreakpoint = Math.min(
    ...latestMinimumEffects.map(({ scope }) => scope.maximumBreakpointIndex),
  );
  const highestPriorityStates = new Set(
    latestMinimumEffects
      .filter(({ scope }) => scope.maximumBreakpointIndex === minimumMaximumBreakpoint)
      .map(({ effect }) => effect.isVisible),
  );
  return highestPriorityStates.size === 1
    ? (highestPriorityStates.values().next().value ?? null)
    : null;
};

export const getTailwindVisibilityAtBreakpoints = (
  className: string,
): ReadonlyArray<boolean> | null => {
  const scopedEffects: TailwindScopedVisibilityEffect[] = [];
  for (const token of splitTailwindClassName(className).map(parseTailwindClassNameToken)) {
    const resolution = getTailwindVisibilityEffect(token.utility);
    if (resolution.status === "not-relevant") continue;
    const scope = getResponsiveVariantScope(token.variants);
    if (scope === null) return null;
    if (!scope || scope.minimumBreakpointIndex >= scope.maximumBreakpointIndex) continue;
    if (
      resolution.status === "unknown" ||
      resolution.propertyName === null ||
      resolution.isVisible === null
    ) {
      return null;
    }
    const effect: TailwindVisibilityEffect = {
      isVisible: resolution.isVisible,
      propertyName: resolution.propertyName,
    };
    scopedEffects.push({ effect, scope, token });
  }

  const visibilityAtBreakpoints: boolean[] = [];
  for (
    let breakpointIndex = 0;
    breakpointIndex < TAILWIND_BREAKPOINT_NAMES.length;
    breakpointIndex += 1
  ) {
    const displayVisibility = resolveVisibilityProperty(scopedEffects, breakpointIndex, "display");
    const visibilityVisibility = resolveVisibilityProperty(
      scopedEffects,
      breakpointIndex,
      "visibility",
    );
    if (displayVisibility === null || visibilityVisibility === null) return null;
    visibilityAtBreakpoints.push(displayVisibility && visibilityVisibility);
  }
  return visibilityAtBreakpoints;
};

File: native/oxlint/rules/get-tailwind-visibility-at-breakpoints.rs
const TAILWIND_BREAKPOINT_NAMES: [&str; 6] = ["", "sm", "md", "lg", "xl", "2xl"];

#[derive(Clone, Copy)]
struct TailwindResponsiveVariantScope {
    maximum_breakpoint_index: usize,
    minimum_breakpoint_index: usize,
    specificity: usize,
}

#[derive(Clone, Copy)]
struct TailwindScopedVisibilityEffect {
    effect: TailwindVisibilityEffect,
    is_important: bool,
    scope: TailwindResponsiveVariantScope,
}

fn get_tailwind_visibility_at_breakpoints(class_name: &str) -> Option<Vec<bool>> {
    let mut scoped_effects = Vec::new();
    for token in tailwind_class_name_tokens(class_name) {
        let effect = match get_tailwind_visibility_effect(token.utility) {
            TailwindVisibilityEffectResolution::Known(effect) => effect,
            TailwindVisibilityEffectResolution::NotRelevant => continue,
            TailwindVisibilityEffectResolution::Unknown => return None,
        };
        let scope = get_responsive_variant_scope(&token.variants)?;
        if scope.minimum_breakpoint_index >= scope.maximum_breakpoint_index {
            continue;
        }
        scoped_effects.push(TailwindScopedVisibilityEffect {
            effect,
            is_important: token.is_important,
            scope,
        });
    }

    let mut visibility_at_breakpoints = Vec::with_capacity(TAILWIND_BREAKPOINT_NAMES.len());
    for breakpoint_index in 0..TAILWIND_BREAKPOINT_NAMES.len() {
        let display_visibility = resolve_visibility_property(
            &scoped_effects,
            breakpoint_index,
            TailwindVisibilityProperty::Display,
        )?;
        let visibility_visibility = resolve_visibility_property(
            &scoped_effects,
            breakpoint_index,
            TailwindVisibilityProperty::Visibility,
        )?;
        visibility_at_breakpoints.push(display_visibility && visibility_visibility);
    }
    Some(visibility_at_breakpoints)
}

fn get_responsive_variant_scope(variants: &[&str]) -> Option<TailwindResponsiveVariantScope> {
    let mut minimum_breakpoint_index = 0;
    let mut maximum_breakpoint_index = TAILWIND_BREAKPOINT_NAMES.len();
    for variant in variants {
        if let Some(minimum_variant_index) = get_tailwind_breakpoint_index(variant)
            && minimum_variant_index > 0
        {
            minimum_breakpoint_index = minimum_breakpoint_index.max(minimum_variant_index);
            continue;
        }
        if let Some(maximum_variant) = variant.strip_prefix("max-")
            && let Some(maximum_variant_index) = get_tailwind_breakpoint_index(maximum_variant)
            && maximum_variant_index > 0
        {
            maximum_breakpoint_index = maximum_breakpoint_index.min(maximum_variant_index);
            continue;
        }
        return None;
    }
    Some(TailwindResponsiveVariantScope {
        maximum_breakpoint_index,
        minimum_breakpoint_index,
        specificity: variants.len(),
    })
}

fn get_tailwind_breakpoint_index(breakpoint_name: &str) -> Option<usize> {
    TAILWIND_BREAKPOINT_NAMES
        .iter()
        .position(|candidate| *candidate == breakpoint_name)
}

fn resolve_visibility_property(
    scoped_effects: &[TailwindScopedVisibilityEffect],
    breakpoint_index: usize,
    property: TailwindVisibilityProperty,
) -> Option<bool> {
    let is_applicable = |scoped_effect: &&TailwindScopedVisibilityEffect| {
        scoped_effect.effect.property == property
            && breakpoint_index >= scoped_effect.scope.minimum_breakpoint_index
            && breakpoint_index < scoped_effect.scope.maximum_breakpoint_index
    };
    let applicable_effects = scoped_effects
        .iter()
        .filter(is_applicable)
        .collect::<Vec<_>>();
    if applicable_effects.is_empty() {
        return Some(true);
    }
    let has_important_effect = applicable_effects.iter().any(|effect| effect.is_important);
    let highest_importance_effects = applicable_effects
        .into_iter()
        .filter(|effect| !has_important_effect || effect.is_important)
        .collect::<Vec<_>>();
    let maximum_specificity = highest_importance_effects
        .iter()
        .map(|effect| effect.scope.specificity)
        .max()?;
    let highest_specificity_effects = highest_importance_effects
        .into_iter()
        .filter(|effect| effect.scope.specificity == maximum_specificity)
        .collect::<Vec<_>>();
    let maximum_minimum_breakpoint = highest_specificity_effects
        .iter()
        .map(|effect| effect.scope.minimum_breakpoint_index)
        .max()?;
    let latest_minimum_effects = highest_specificity_effects
        .into_iter()
        .filter(|effect| effect.scope.minimum_breakpoint_index == maximum_minimum_breakpoint)
        .collect::<Vec<_>>();
    let minimum_maximum_breakpoint = latest_minimum_effects
        .iter()
        .map(|effect| effect.scope.maximum_breakpoint_index)
        .min()?;
    let mut resolved_visibility = None;
    for effect in latest_minimum_effects
        .into_iter()
        .filter(|effect| effect.scope.maximum_breakpoint_index == minimum_maximum_breakpoint)
    {
        if resolved_visibility.is_some_and(|is_visible| is_visible != effect.effect.is_visible) {
            return None;
        }
        resolved_visibility = Some(effect.effect.is_visible);
    }
    resolved_visibility
}
